vllm/compilation/pass_manager.py

```diff
@@ -0,0 +1,88 @@
+from typing import List
+
+from torch import fx as fx
+from torch._inductor.codecache import BypassFxGraphCache
+
+from vllm.config import CompilationConfig
+from vllm.logger import init_logger
+
+from .fix_functionalization import FixFunctionalizationPass
+from .fusion import FusionPass
+from .inductor_pass import InductorPass, InductorPassType
+from .reshapes import RedundantReshapesPass
+
+logger = init_logger(__name__)
+
+
+class PostGradPassManager:
+    """
+    The pass manager for post-grad passes.
+    It handles configuration, adding custom passes, and running passes.
+    It also supports pickling, which is used by the Inductor code cache.
+    TODO(torch==2.6), use CustomGraphPass
+    (torch._inductor.custom_graph_pass.CustomGraphPass)
+
+    The order of the post-grad post-passes is:
+    1. passes (constructor parameter)
+    2. default passes (RedundantReshapesPass, FusionPass)
+    3. config["post_grad_custom_post_pass"] (if it exists)
+    4. fix_functionalization
+    This way, all passes operate on a functionalized graph.
+    """
+
+    def __init__(self):
+        self.passes: List[InductorPassType] = []
+
+    def __call__(self, graph: fx.Graph):
+        for pass_ in self.passes:
+            pass_(graph)
+
+        # always run fix_functionalization last
+        self.fix_functionalization(graph)
+
+    def configure(self, pass_config: CompilationConfig.PassConfig):
+        self.pass_config = pass_config
+        if pass_config.enable_reshape:
+            self.passes += [RedundantReshapesPass(pass_config)]
+
+        if pass_config.enable_fusion:
+            self.passes += [FusionPass.instance(pass_config)]
+
+        self.fix_functionalization = FixFunctionalizationPass(pass_config)
+
+    def add(self, pass_: InductorPassType):
+        self.passes.append(pass_)
+
+    def __getstate__(self):
+        """
+        Custom pickling for the pass manager, as some passes cannot be pickled.
+        Pickling occurs because the pass manager is set as the value of
+        `config["post_grad_custom_post_pass"]` in the Inductor config.
+
+        TODO(torch==2.6), use the `uuid` method in CustomGraphPass instead.
+        """
+        state = {"pass_config": self.pass_config}
+        passes_state = []
+        for pass_ in self.passes:
+            if isinstance(pass_, InductorPass):
+                passes_state.append(pass_.uuid())
```

Comment:
with the discussion in the channel, I agree it is useful to cache based on the content of the pass.

however, i still think using the hash of the whole file is too overkill. you can use python `inspect` module to get the source code from the function object, which i think should be enough.

you cannot really cache based on the full code, e.g. even if the file does not change, a dependent package changes, and you cannot invalidate the cache because you don't know dependent package version changes.

Reply:
even if you want to hash the whole file, i think python's `inspect` module should give your enough information.

my request is, if a developer passes a function's qualname, it should just work. they don't need to know this uuid stuff.

Reply:
Okay yeah that's fair. I'll look into `python.inspect`. Should we use this for vLLM passes as well? Or just custom user passes?

Reply:
The nice things about the current approach with vLLM passes is that we do know the files a pass depends on. But I think inspect would be fine too.

Reply:
we can inspect both vllm passes and custom user passes.

Reply:
Done